# Projeto de Bases de Dados - Parte 2

### Grupo 25
<dl>
    <dt>26 horas (33.3%)</dt>
    <dd>ist1109474 Daniel Borges</dd>
    <dt>26 horas (33.3%)</dt>
    <dd>ist1110239 André Pagaime</dd>
    <dt>26 horas (33.3%)</dt>
    <dd>ist1110181 Duarte Cruz</dd>
<dl>

In [10]:
%load_ext sql
%config SqlMagic.displaycon = 0
%config SqlMagic.displaylimit = None
%sql postgresql+psycopg://proj:proj@postgres/proj

The sql extension is already loaded. To reload it, use:
  %reload_ext sql


displaylimit: Value None will be treated as 0 (no limit)

Connecting and switching to connection 'postgresql+psycopg://proj:***@postgres/proj'

## 0. Carregamento da Base de Dados

Crie a base de dados “Aviacao” no PostgreSQL e execute os comandos para criação das tabelas desta base de dados apresentados no ficheiro “aviacao.sql”

## 1. Restrições de Integridade [3 valores]

Implemente na base de dados “Aviacao” as seguintes restrições de integridade, podendo recorrer a Triggers caso estritamente necessário:

(RI-1) Aquando do check-in (i.e. quando se define o assento em bilhete) a classe do bilhete tem de corresponder à classe do assento e o aviao do assento tem de corresponder ao aviao do voo

In [14]:
%%sql
-- (RI-1)

CREATE OR REPLACE FUNCTION checkin_verifications() RETURNS TRIGGER AS
$$
DECLARE
    classe_assento BOOLEAN;
    aviao_voo VARCHAR(80);
BEGIN
    IF NEW.lugar IS NOT NULL AND NEW.no_serie IS NOT NULL AND (NEW.lugar IS DISTINCT FROM OLD.lugar OR NEW.no_serie IS DISTINCT FROM OLD.no_serie) THEN
        SELECT prim_classe INTO classe_assento
        FROM assento
        WHERE lugar = NEW.lugar AND no_serie = NEW.no_serie;

        IF NEW.prim_classe IS DISTINCT FROM classe_assento THEN
            RAISE EXCEPTION 'A classe do bilhete (primeira classe: %) para o bilhete ID % não corresponde à classe do assento (primeira classe: %).', NEW.prim_classe, NEW.id, classe_assento;
        END IF;

        SELECT no_serie INTO aviao_voo
        FROM voo
        WHERE id = NEW.voo_id;

        IF aviao_voo IS DISTINCT FROM NEW.no_serie THEN
            RAISE EXCEPTION 'O avião do assento (número de série: %) para o bilhete ID % não corresponde ao avião do voo (número de série: %).', NEW.no_serie, NEW.id, aviao_voo;
        END IF;
    END IF;

    RETURN NEW;
END;
$$ LANGUAGE plpgsql;

CREATE OR REPLACE TRIGGER checkin_trigger BEFORE INSERT OR UPDATE ON bilhete
FOR EACH ROW EXECUTE FUNCTION checkin_verifications();

++
||
++
++

(RI-2) O número de bilhetes de cada classe vendidos para cada voo não pode exceder a capacidade (i.e., número de assentos) do avião para essa classe

In [15]:
%%sql
-- (RI-2)

CREATE OR REPLACE FUNCTION class_verification() RETURNS TRIGGER AS
$$
DECLARE
    capacidade INTEGER;
    vendidos INTEGER;
    aviao_voo VARCHAR(80);
    nome_classe VARCHAR(20);
BEGIN
    SELECT no_serie INTO aviao_voo
    FROM voo
    WHERE id = NEW.voo_id;

    SELECT COUNT(*) INTO capacidade
    FROM assento
    WHERE no_serie = aviao_voo AND prim_classe = NEW.prim_classe;

    SELECT COUNT(*) INTO vendidos
    FROM bilhete
    WHERE voo_id = NEW.voo_id AND prim_classe = NEW.prim_classe;

    IF vendidos >= capacidade THEN
        IF NEW.prim_classe THEN
            nome_classe := 'Primeira Classe';
        ELSE
            nome_classe := 'Segunda Classe';
        END IF;

        RAISE EXCEPTION 'Capacidade da % para o voo ID % a ser excedida.', nome_classe, NEW.voo_id;
    END IF;

    RETURN NEW;
END;
$$ LANGUAGE plpgsql;

CREATE OR REPLACE TRIGGER class_trigger BEFORE INSERT ON bilhete
FOR EACH ROW EXECUTE FUNCTION class_verification();

++
||
++
++

(RI-3) A hora da venda tem de ser anterior à hora de partida de todos os voos para os quais foram comprados bilhetes na venda

In [16]:
%%sql
-- (RI-3)

CREATE OR REPLACE FUNCTION ticket_verification() RETURNS TRIGGER AS
$$
DECLARE
    hora_venda TIMESTAMP;
    hora_voo TIMESTAMP;
BEGIN
    SELECT hora INTO hora_venda
    FROM venda
    WHERE codigo_reserva = NEW.codigo_reserva;

    SELECT hora_partida INTO hora_voo
    FROM voo
    WHERE id = NEW.voo_id;

    IF hora_venda >= hora_voo THEN
        RAISE EXCEPTION 'A hora da venda (%) para o bilhete ID % (Venda: %, Voo: %) não pode ser igual ou posterior à hora de partida do voo (%).', hora_venda, NEW.id, NEW.codigo_reserva, NEW.voo_id, hora_voo;
    END IF;

    RETURN NEW;
END;
$$ LANGUAGE plpgsql;

CREATE OR REPLACE TRIGGER ticket_trigger BEFORE INSERT ON bilhete
FOR EACH ROW EXECUTE FUNCTION ticket_verification();

++
||
++
++

## 2. Preenchimento da Base de Dados [2 valores]

Preencha todas as tabelas da base de dados de forma consistente (após execução do ponto anterior) com os seguintes requisitos adicionais de cobertura:
- ≥10 aeroportos internacionais (reais) localizados na Europa, com pelo menos 2 cidades tendo 2 aeroportos
- ≥10 aviões de ≥3 modelos distintos (reais), com um número de assentos realista; assuma que as primeiras ~10% filas são de 1a classe
- ≥5 voos por dia entre 1 de Janeiro e 31 de Julho de 2025, cobrindo todos os aeroportos e todos os aviões; garanta que para cada voo entre dois aeroportos se segue um voo no sentido oposto; garanta ainda que cada avião tem partida no aeroporto da sua chegada anterior
- ≥30.000 bilhetes vendidos até à data presente, correspondendo a ≥10.000 vendas, com todo os bilhetes de voos já realizados tendo feito check-in, e com todos os voos tendo bilhetes de primeira e segunda classe vendidos
Deve ainda garantir que todas as consultas necessárias para a realização dos pontos seguintes do projeto produzem um resultado não vazio.

O código para preenchimento da base de dados deve ser compilado num ficheiro "populate.sql", anexado ao relatório, que contém com comandos INSERT ou alternativamente comandos COPY que populam as tabelas a partir de ficheiros de texto, também eles anexados ao relatório.

## 3. Desenvolvimento de Aplicação [5 valores]

Crie um protótipo de RESTful web service para gestão de consultas por acesso programático à base de dados ‘Aviacao’ através de uma API que devolve respostas em JSON, implementando os seguintes endpoints REST:

|Endpoint|Descrição|
|--------|---------|
|/|Lista todos os aeroportos (nome e cidade).|
|/voos/\<partida>/|Lista todos os voos (número de série do avião,  hora de partida e aeroporto de chegada) que partem do aeroporto de \<partida> até 12h após o momento da consulta.|
|/voos/\<partida>/\<chegada>/|Lista os próximos três voos (número de série do avião e hora de partida) entre o aeroporto de \<partida> e o aeroporto de \<chegada> para os quais ainda há bilhetes disponíveis.|
|/compra/\<voo>/|Faz uma compra de um ou mais bilhetes para o \<voo>, populando as tabelas \<venda> e \<bilhete>. Recebe como argumentos o nif do cliente, e uma lista de pares (nome de passageiro, classe de bilhete) especificando os bilhetes a comprar.|
|/checkin/\<bilhete>/|Faz o check-in de um bilhete, atribuindo-lhe automaticamente um assento da classe correspondente.|

## 3. Vistas [2 valores]

Crie uma vista materializada que detalhe as informações mais importantes sobre os voos, combinando a informação de várias tabelas da base de dados. A vista deve ter o seguinte esquema:

 *estatisticas_voos(no_serie, hora_partida, cidade_partida, pais_partida, cidade_chegada, pais_chegada, ano, mes, dia_do_mes, dia_da_semana, passageiros_1c, passageiros_2c, assentos_1c, assentos_2c, vendas_1c, vendas_2c)*

em que:
- *no_serie, hora_partida*: correspondem aos atributos homónimos da tabela *voo*
- *cidade_partida, pais_partida, cidade_chegada, pais_chegada*: correspondem aos atributos *cidade* e *pais* da tabela *aeroporto*, para o aeroporto de *partida* e *chegada* do *voo*
- *ano, mes, dia_do_mes* e *dia_da_semana*: são derivados do atributo *hora_partida* da tabela *voo*
- *passageiros_1c, passageiros_2c:*: correspondem ao número total de bilhetes vendidos para o voo, de primeira e segunda classe respectivamente
- *assentos_1c, assentos_2c:*: correspondem ao número de assentos de primeira e segunda classe no avião que realiza o voo
- *vendas_1c, vendas_2c*: correspondem ao somatório total dos preços dos bilhetes vendidos para o voo, de primeira e segunda classe respectivamente

In [6]:
%%sql
DROP MATERIALIZED VIEW IF EXISTS estatisticas_voos;

CREATE MATERIALIZED VIEW estatisticas_voos AS
WITH AssentosPorAviao AS (
    SELECT
        no_serie,
        SUM(CASE WHEN prim_classe = TRUE THEN 1 ELSE 0 END) AS total_assentos_1c,
        SUM(CASE WHEN prim_classe = FALSE THEN 1 ELSE 0 END) AS total_assentos_2c
    FROM
        assento
    GROUP BY
        no_serie
),
BilhetesEVendasPorVoo AS (
    SELECT
        voo_id,
        COUNT(CASE WHEN prim_classe = TRUE THEN 1 END) AS passageiros_1c_agg,
        COUNT(CASE WHEN prim_classe = FALSE THEN 1 END) AS passageiros_2c_agg,
        SUM(CASE WHEN prim_classe = TRUE THEN preco ELSE 0 END) AS vendas_1c_agg,
        SUM(CASE WHEN prim_classe = FALSE THEN preco ELSE 0 END) AS vendas_2c_agg
    FROM
        bilhete
    GROUP BY
        voo_id
)
SELECT
    voo.no_serie,
    voo.hora_partida,
    aeroportoPartida.cidade AS cidade_partida,
    aeroportoPartida.pais AS pais_partida,
    aeroportoChegada.cidade AS cidade_chegada,
    aeroportoChegada.pais AS pais_chegada,
    EXTRACT(YEAR FROM voo.hora_partida) AS ano,
    EXTRACT(MONTH FROM voo.hora_partida) AS mes,
    EXTRACT(DAY FROM voo.hora_partida) AS dia_do_mes,
    EXTRACT(DOW FROM voo.hora_partida) AS dia_da_semana,

    COALESCE(bilhetesVoo.passageiros_1c_agg, 0) AS passageiros_1c,
    COALESCE(bilhetesVoo.passageiros_2c_agg, 0) AS passageiros_2c,

    COALESCE(assentosAviao.total_assentos_1c, 0) AS assentos_1c,
    COALESCE(assentosAviao.total_assentos_2c, 0) AS assentos_2c,

    COALESCE(bilhetesVoo.vendas_1c_agg, 0) AS vendas_1c,
    COALESCE(bilhetesVoo.vendas_2c_agg, 0) AS vendas_2c
    
FROM
    voo
JOIN
    aeroporto aeroportoPartida ON voo.partida = aeroportoPartida.codigo 
JOIN
    aeroporto aeroportoChegada ON voo.chegada = aeroportoChegada.codigo
LEFT JOIN
    BilhetesEVendasPorVoo bilhetesVoo ON voo.id = bilhetesVoo.voo_id
LEFT JOIN
    AssentosPorAviao assentosAviao ON voo.no_serie = assentosAviao.no_serie
ORDER BY
    voo.hora_partida DESC;

1736 rows affected.

++
||
++
++

## 5. Análise de Dados SQL e OLAP [5 valores]

Usando apenas a vista *estatisticas_voos* desenvolvida no ponto anterior, e **sem recurso ao operador LIMIT e com recurso ao operador WITH apenas se estritamente necessário**, apresente a consulta SQL mais sucinta para cada um dos seguintes objetivos analíticos da empresa. Pode usar agregações OLAP para os objetivos em que lhe parecer adequado.

1. Determinar a(s) rota(s) que tem/têm a maior procura para efeitos de aumentar a frequência de voos dessa(s) rota(s). Entende-se por rota um trajeto aéreo entre quaisquer duas cidades,  independentemente do sentido (e.g., voos Lisboa-Paris e Paris-Lisboa contam para a mesma rota). Considera-se como indicador da procura de uma rota o preenchimento médio dos aviões (i.e., o rácio entre o número total de passageiros e a capacidade total do avião) no último ano.

In [16]:
%%sql

SELECT
    sub.rota
FROM (
    SELECT
        CASE
            WHEN voo.cidade_partida < voo.cidade_chegada OR (voo.cidade_partida = voo.cidade_chegada AND voo.pais_partida < voo.pais_chegada) THEN
                voo.cidade_partida || ', ' || voo.pais_partida || ' - ' || voo.cidade_chegada || ', ' || voo.pais_chegada
            ELSE
                voo.cidade_chegada || ', ' || voo.pais_chegada || ' - ' || voo.cidade_partida || ', ' || voo.pais_partida
        END AS rota,
        DENSE_RANK() OVER (ORDER BY AVG(CAST((voo.passageiros_1c + voo.passageiros_2c) AS NUMERIC) / (voo.assentos_1c + voo.assentos_2c)) DESC) AS rank_preenchimento
    FROM
        estatisticas_voos voo
    WHERE
        (voo.assentos_1c + voo.assentos_2c) > 0
        AND voo.hora_partida >= (CURRENT_TIMESTAMP - INTERVAL '1 year')
    GROUP BY
        rota
) AS sub
WHERE
    sub.rank_preenchimento = 1;

1 rows affected.

+-----------------------------------------------+
|                      rota                     |
+-----------------------------------------------+
| Amesterdão, Países Baixos - Bruxelas, Bélgica |
+-----------------------------------------------+

2. Determinar as rotas pelas quais nos últimos 3 meses passaram todos os aviões da empresa, para efeitos de melhorar a gestão da frota.

In [8]:
%%sql

SELECT
    CASE
        WHEN voo1.cidade_partida < voo1.cidade_chegada OR (voo1.cidade_partida = voo1.cidade_chegada AND voo1.pais_partida < voo1.pais_chegada) THEN
            voo1.cidade_partida || ', ' || voo1.pais_partida || ' - ' || voo1.cidade_chegada || ', ' || voo1.pais_chegada
        ELSE
            voo1.cidade_chegada || ', ' || voo1.pais_chegada || ' - ' || voo1.cidade_partida || ', ' || voo1.pais_partida
    END AS rota
FROM
    estatisticas_voos voo1
WHERE
    voo1.hora_partida >= (CURRENT_TIMESTAMP - INTERVAL '3 months')
GROUP BY
    rota
HAVING
    COUNT(DISTINCT voo1.no_serie) = (
        SELECT
            COUNT(DISTINCT voo2.no_serie)
        FROM
            estatisticas_voos voo2
    );

1 rows affected.

+------------------------------------------+
|                   rota                   |
+------------------------------------------+
| Londres, Reino Unido - Munique, Alemanha |
+------------------------------------------+

3. Explorar a rentabilidade da empresa (vendas globais e por classe) nas dimensões espaço (global > pais > cidade, para a partida e chegada em simultâneo) e tempo (global > ano > mes > dia_do_mes), como apoio a um relatório executivo.

In [45]:
%%sql

SELECT
    CASE WHEN GROUPING(pais_partida) = 0 THEN pais_partida ELSE 'Global' END AS país_de_partida,
    CASE WHEN GROUPING(cidade_partida) = 0 THEN cidade_partida ELSE '-' END AS cidade_de_partida,
    CASE WHEN GROUPING(pais_chegada) = 0 THEN pais_chegada ELSE 'Global' END AS país_de_chegada,
    CASE WHEN GROUPING(cidade_chegada) = 0 THEN cidade_chegada ELSE '-' END AS cidade_de_chegada,

    CASE WHEN GROUPING(ano) = 0 THEN ano::text ELSE 'Global' END AS ano,
    CASE WHEN GROUPING(mes) = 0 THEN mes::text ELSE '-' END AS mês,
    CASE WHEN GROUPING(dia_do_mes) = 0 THEN dia_do_mes::text ELSE '-' END AS dia,
    
    SUM(vendas_2c) AS vendas_2_classe,
    SUM(vendas_1c) AS vendas_1_classe,
    SUM(vendas_1c + vendas_2c) AS vendas_globais
FROM
    estatisticas_voos
GROUP BY GROUPING SETS (
    (),
    (ano),
    (ano, mes),
    (ano, mes, dia_do_mes),

    (pais_partida, pais_chegada),
    (pais_partida, pais_chegada, ano),
    (pais_partida, pais_chegada, ano, mes),
    (pais_partida, pais_chegada, ano, mes, dia_do_mes),

    (pais_partida, pais_chegada, cidade_partida, cidade_chegada),
    (pais_partida, pais_chegada, cidade_partida, cidade_chegada, ano),
    (pais_partida, pais_chegada, cidade_partida, cidade_chegada, ano, mes),
    (pais_partida, pais_chegada, cidade_partida, cidade_chegada, ano, mes, dia_do_mes)
)
ORDER BY
    GROUPING(pais_partida) DESC, pais_partida,
    GROUPING(pais_chegada) DESC, pais_chegada,
    GROUPING(cidade_partida) DESC, cidade_partida,
    GROUPING(cidade_chegada) DESC, cidade_chegada,
    
    GROUPING(ano) DESC, ano DESC,
    GROUPING(mes) DESC, mes DESC,
    GROUPING(dia_do_mes) DESC, dia_do_mes DESC;

6193 rows affected.

+-----------------+-------------------+-----------------+-------------------+--------+-----+-----+-----------------+-----------------+----------------+
| país_de_partida | cidade_de_partida | país_de_chegada | cidade_de_chegada |  ano   | mês | dia | vendas_2_classe | vendas_1_classe | vendas_globais |
+-----------------+-------------------+-----------------+-------------------+--------+-----+-----+-----------------+-----------------+----------------+
|      Global     |         -         |      Global     |         -         | Global |  -  |  -  |   18499229.75   |    3750797.00   |  22250026.75   |
|      Global     |         -         |      Global     |         -         |  2025  |  -  |  -  |   18499229.75   |    3750797.00   |  22250026.75   |
|      Global     |         -         |      Global     |         -         |  2025  |  7  |  -  |    2618749.77   |    536505.00    |   3155254.77   |
|      Global     |         -         |      Global     |         -         |  2025  |  7  |  31 |     88580.18    |     16607.00    |   105187.18    |
|      Global     |         -         |      Global     |         -         |  2025  |  7  |  30 |     65832.36    |     10774.00    |    76606.36    |
|      Global     |         -         |      Global     |         -         |  2025  |  7  |  29 |    103356.64    |     22332.00    |   125688.64    |
|      Global     |         -         |      Global     |         -         |  2025  |  7  |  28 |    100300.04    |     22026.00    |   122326.04    |
|      Global     |         -         |      Global     |         -         |  2025  |  7  |  27 |     85328.62    |     12265.00    |    97593.62    |
|      Global     |         -         |      Global     |         -         |  2025  |  7  |  26 |    115972.93    |     22814.00    |   138786.93    |
|      Global     |         -         |      Global     |         -         |  2025  |  7  |  25 |     52234.19    |     11692.00    |    63926.19    |
|      Global     |         -         |      Global     |         -         |  2025  |  7  |  24 |     52732.09    |     12393.00    |    65125.09    |
|      Global     |         -         |      Global     |         -         |  2025  |  7  |  23 |     54374.10    |     11445.00    |    65819.10    |
|      Global     |         -         |      Global     |         -         |  2025  |  7  |  22 |     77942.88    |     16000.00    |    93942.88    |
|      Global     |         -         |      Global     |         -         |  2025  |  7  |  21 |     70129.07    |     12138.00    |    82267.07    |
|      Global     |         -         |      Global     |         -         |  2025  |  7  |  20 |    143712.93    |     28858.00    |   172570.93    |
|      Global     |         -         |      Global     |         -         |  2025  |  7  |  19 |    104243.82    |     22706.00    |   126949.82    |
|      Global     |         -         |      Global     |         -         |  2025  |  7  |  18 |     97432.67    |     19630.00    |   117062.67    |
|      Global     |         -         |      Global     |         -         |  2025  |  7  |  17 |    149922.96    |     32220.00    |   182142.96    |
|      Global     |         -         |      Global     |         -         |  2025  |  7  |  16 |     98113.21    |     16451.00    |   114564.21    |
|      Global     |         -         |      Global     |         -         |  2025  |  7  |  15 |     87558.71    |     18702.00    |   106260.71    |
|      Global     |         -         |      Global     |         -         |  2025  |  7  |  14 |     59971.61    |     10675.00    |    70646.61    |
|      Global     |         -         |      Global     |         -         |  2025  |  7  |  13 |     56293.14    |     12252.00    |    68545.14    |
|      Global     |         -         |      Global     |         -         |  2025  |  7  |  12 |     64139.35    |     13878.00    |    78017.35    |
|      Global     |         -         |      Glo

4. Descobrir se há algum padrão ao longo da semana no rácio entre passageiros de primeira e segunda classe, com drill down na dimensão espaço (global > pais > cidade), que justifique uma abordagem mais flexível à divisão das classes.

In [62]:
%%sql

SELECT
    CASE WHEN GROUPING(pais_partida) = 0 THEN pais_partida ELSE 'Global' END AS país_de_partida,
    CASE WHEN GROUPING(cidade_partida) = 0 THEN cidade_partida ELSE '-' END AS cidade_de_partida,

    CASE WHEN GROUPING(pais_chegada) = 0 THEN pais_chegada ELSE 'Global' END AS país_de_chegada,
    CASE WHEN GROUPING(cidade_chegada) = 0 THEN cidade_chegada ELSE '-' END AS cidade_de_chegada,

    CASE
        WHEN GROUPING(dia_da_semana) = 1 THEN 'Global'
        WHEN dia_da_semana = 0 THEN 'Domingo'
        WHEN dia_da_semana = 1 THEN 'Segunda-feira'
        WHEN dia_da_semana = 2 THEN 'Terça-feira'
        WHEN dia_da_semana = 3 THEN 'Quarta-feira'
        WHEN dia_da_semana = 4 THEN 'Quinta-feira'
        WHEN dia_da_semana = 5 THEN 'Sexta-feira'
        WHEN dia_da_semana = 6 THEN 'Sábado'
    END AS nome_dia_da_semana,

    CASE
        WHEN SUM(passageiros_2c) > 0 THEN (SUM(passageiros_1c)::numeric / SUM(passageiros_2c)::numeric)
        ELSE NULL
    END AS rácio_1_classe_com_2_classe
FROM
    estatisticas_voos
GROUP BY GROUPING SETS (
    (),
    (dia_da_semana),

    (pais_partida, pais_chegada),
    (pais_partida, pais_chegada, dia_da_semana),

    (pais_partida, pais_chegada, cidade_partida, cidade_chegada),
    (pais_partida, pais_chegada, cidade_partida, cidade_chegada, dia_da_semana)
)
ORDER BY
    GROUPING(pais_partida) DESC, pais_partida,
    GROUPING(pais_chegada) DESC, pais_chegada,
    GROUPING(cidade_partida) DESC, cidade_partida,
    GROUPING(cidade_chegada) DESC, cidade_chegada,
    
    GROUPING(dia_da_semana) DESC, dia_da_semana ASC;

2273 rows affected.

+-----------------+-------------------+-----------------+-------------------+--------------------+-----------------------------+
| país_de_partida | cidade_de_partida | país_de_chegada | cidade_de_chegada | nome_dia_da_semana | rácio_1_classe_com_2_classe |
+-----------------+-------------------+-----------------+-------------------+--------------------+-----------------------------+
|      Global     |         -         |      Global     |         -         |       Global       |    0.13366808291489635638   |
|      Global     |         -         |      Global     |         -         |      Domingo       |    0.13475066843184591024   |
|      Global     |         -         |      Global     |         -         |   Segunda-feira    |    0.13175385096567362577   |
|      Global     |         -         |      Global     |         -         |    Terça-feira     |    0.13454555933375597948   |
|      Global     |         -         |      Global     |         -         |    Quarta-feira    |    0.13345415602603179834   |
|      Global     |         -         |      Global     |         -         |    Quinta-feira    |    0.13514927759527893253   |
|      Global     |         -         |      Global     |         -         |    Sexta-feira     |    0.12971736644339160268   |
|      Global     |         -         |      Global     |         -         |       Sábado       |    0.13627624963028689737   |
|     Alemanha    |         -         |     Alemanha    |         -         |       Global       |    0.08762886597938144330   |
|     Alemanha    |         -         |     Alemanha    |         -         |    Quarta-feira    |    0.08762886597938144330   |
|     Alemanha    |     Frankfurt     |     Alemanha    |      Munique      |       Global       |    0.08602150537634408602   |
|     Alemanha    |     Frankfurt     |     Alemanha    |      Munique      |    Quarta-feira    |    0.08602150537634408602   |
|     Alemanha    |      Munique      |     Alemanha    |     Frankfurt     |       Global       |    0.08910891089108910891   |
|     Alemanha    |      Munique      |     Alemanha    |     Frankfurt     |    Quarta-feira    |    0.08910891089108910891   |
|     Alemanha    |         -         |     Áustria     |         -         |       Global       |    0.14940098661028893587   |
|     Alemanha    |         -         |     Áustria     |         -         |      Domingo       |    0.17142857142857142857   |
|     Alemanha    |         -         |     Áustria     |         -         |   Segunda-feira    |    0.15699658703071672355   |
|     Alemanha    |         -         |     Áustria     |         -         |    Quarta-feira    |    0.13947990543735224586   |
|     Alemanha    |         -         |     Áustria     |         -         |    Sexta-feira     |    0.16600790513833992095   |
|     Alemanha    |         -         |     Áustria     |         -         |       Sábado       |    0.13225806451612903226   |
|     Alemanha    |     Frankfurt     |     Áustria     |       Viena       |       Global       |    0.14664804469273743017   |
|     Alemanha    |     Frankfurt     |     Áustria     |       Viena       |   Segunda-feira    |    0.15699658703071672355   |
|     Alemanha    |     Frankfurt     |     Áustria     |       Viena       |    Quarta-feira    |    0.13947990543735224586   |
|     Alemanha    |      Munique      |     Áustria     |       Viena       |       Global       |    0.15220483641536273115   |
|     Alemanha    |      Munique      |     Áustria     |       Viena       |      Domingo       |    0.17142857142857142857   |
|     Alemanha    |      Munique      |     Áustria     |       Viena       |    Sexta-feira     |    0.16600790513833992095   |
|     Alemanha    |      Munique      |     Áustria     |       Viena       |       Sábado       |    0.13225806451612903226   |
|     Alemanha    |         -         |     Bélgica     |         -         |       Global       |    0.13639788997739261492   |
|

## 6. Índices [3 valores]

É expectável que seja necessário executar consultas semelhantes ao colectivo das consultas do ponto anterior diversas vezes ao longo do tempo, e pretendemos otimizar o desempenho da vista estatisticas_voos para esse efeito. Crie sobre a vista o(s) índice(s) que achar mais indicados para fazer essa otimização, justificando a sua escolha com argumentos teóricos e com demonstração prática do ganho em eficiência do índice por meio do comando EXPLAIN ANALYSE. Deve procurar uma otimização coletiva das consultas, evitando criar índices excessivos, particularmente se estes trazem apenas ganhos incrementais a uma das consultas.

Código para criação dos índices

In [ ]:
%%sql

DROP INDEX IF EXISTS idx_hora_partida;
DROP INDEX IF EXISTS idx_no_serie;
DROP INDEX IF EXISTS idx_localizacao;

CREATE INDEX idx_hora_partida ON estatisticas_voos (hora_partida);
CREATE INDEX idx_no_serie ON estatisticas_voos (no_serie);
CREATE INDEX idx_localizacao ON estatisticas_voos (pais_partida, pais_chegada, cidade_partida, cidade_chegada);

Justificação teórica e prática (sumarizando observações com EXPLAIN ANALYSE)

O índice idx_hora_partida, do tipo B-Tree, foi criado com o objetivo de otimizar a performance de consultas que filtram temporalmente a vista materializada estatisticas_voos. A coluna hora_partida é frequentemente utilizada em cláusulas WHERE para restringir os resultados a um determinado intervalo de tempo. Com este índice evita-se uma leitura sequencial da tabela, o que permite uma filtragem mais eficiente quando existe uma grande quantidade de dados.
Na consulta 5.1, é aplicada um WHERE que seleciona apenas os voos realizados no último ano. Todos os dados na base de dados estão contidos no entre as datas 01-01-2025 e 31-07-2025. Sendo assim, todas as linhas da estatisticas_voos são selecionadas. Tendo esta consulta uma seletividade nula, o índice não é útil, tendo em conta a quantidade de dados que temos (cerca de 1700 voos).
Na consulta 5.2 também existe uma filtragem por tempo parecida com a consulta anterior, sendo de três meses em vez de um ano. Apesar do intervalo temporal mais pequeno, ainda não é pequeno o suficiente para se justificar a utilização do índice considerando o volume de dados que temos. Cerca de metade dos voos na estatisticas_voos aconteceram nos últimos três meses, logo metade das linhas são relevantes, fazendo com que esta consulta tenha uma seletividade baixa, fazendo com o índice que seja menos útil.
Sendo assim, o índice idx_hora_partida é útil quando existe um grande número de voos na estatisticas_voos, o que não é o nosso caso. Devido à baixa quantidade de dados, as consultas têm uma baixa seletividade, levando o query planner a realizar sequential scans em vez de utilizar o índice, uma vez que os benefícios do uso do índice são pouco significativos neste contexto.

O índice idx_no_serie, do tipo B-Tree, foi criado para otimizar consultas que filtram os voos da vista materializada estatisticas_voos por avião. A coluna no_serie pode ser utilizada para filtrar voos pelo avião, sendo útil para isso este índice para evitar uma leitura sequencial da tabela, otimizando as consultas.
Mesmo com poucos voos na vista materializada estatisticas_voos, este índice pode ser útil para otimizar certas operações.
Um exemplo disso está na consulta 5.2, onde existe um COUNT(DISTINCT voo1.no_serie). Sem índice, para realizar essa agregação, o query planner teria que ordenar as linhas para depois fazer uma varredura completa da tabela para fazer a contagem dos elementos distintos da coluna no_serie, o que pode ser computacionalmente caro. Como o índice idx_no_serie já armazena os valores de no_serie ordenados, devido à sua estrutura B-Tree, este pode ser aproveitado para executar essa parte da consulta de forma mais eficiente, permitindo um index only scan, que evita a ordenação e leitura completa da tabela, acelerando o processo de contagem dos valores distintos.
Com este índice e o anterior, a consulta 5.2, apesar de por padrão ser feita sem índices devido ao reduzido número de voos na estatisticas_voos, quando pedimos explicitamente ao query planner que utilize os índices, observamos uma ligeira melhoria do tempo de execução. Sem os índices, obtemos um tempo de execução de 3.329ms (0.448ms de planning + 2.881ms de execution), com os índices obtemos um tempo de execução de 2.423ms (0.252ms de planning + 2.171ms de execution), o que corresponde a uma melhoria de cerca de 27%.

O índice composto idx_localizacao, do tipo B-Tree, foi criado para melhorar o desempenho de consultas à vista materializada estatisticas_voos que utilizam a localização dos aeroportos de partida e chegada.
Dois exemplos de consultas que utilizam a localização são as 5.3 e a 5.4. Em ambas as consultas temos múltiplas agregações, e como vimos acima, índices podem otimizar agregações como as presentes nestas consultas, evitando ordenações e leituras sequênciais na presença de um grande volume de dados. Devido ao número reduzido de voos na vista materializada estatisticas_voos, o query planner utiliza sequential scanning em vez de fazer um index scan com o índice idx_localizacoes. Quando pedido explicitamente a utilização dos índices, o impacto deste índice na performance das consultas 5.3 e 5.4 não é significativo.